### PR TITLE
Create or deprecate a category : Fix #289 #301 

### LINKED ISSUE #289: [Reference Data] as a user i want to create or deprecate a category
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the COMET-WEB
- [x] I have searched [open](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

- [x] create a "form" to provide the properties to define a category (shortname, name, library, is abstract, is deprecated, selection of permissible classes, super categories)
- [x] when the user clicks on the "add category" button on the categories page, show the form to populate with the required data
- [x] when the user clicks on the cancel button, the creation is cancelled
- [x] when the ok button is clicked, a transaction is created and the category is created
- [x] when the deprecate/undeprecate button is clicked, a transaction is created and the category is updated
- [x] add unit tests to view-model (nunit) and view (bunit)

### Logs

  - [ ] There are no logs regarding the issue
  - { ] I have attached a log file or included the relevant section in this issue  
  
### System Configuration

- COMET version:
  - [ ] COMET-WEB: 
  - [ ] Other: 
- Web Services:
  - [ ] COMET Web Services version:

### PR DESCRIPTION
**Pull request by nabilatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-WEB [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WEB-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description

Fix #289 #301
Feature :

- Create or deprecate a category
- Separate permission access rights to write to Site Reference Data library and Model Reference Data library for Categories and ParameterTypes

### CODE REVIEW

File: COMETwebapp.Tests/Components/ReferenceData/CategoriesTableTestFixture.cs [hunk 1]
```diff
@@ -317,5 +318,97 @@ public async Task VerifyOnInitialized()
...
+            this.viewModel.SelectedPermissibleClasses = new List<ClassKindWrapper>() { new ClassKindWrapper(ClassKind.ElementDefinition) };
+
+            await this.viewModel.AddingCategory();
+            CDPMessageBus.Current.SendMessage(new ObjectChangedEvent(this.viewModel.Category, EventKind.Added));
+
+            Assert.Multiple(() => { Assert.That(this.viewModel.Rows.Count, Is.EqualTo(2)); });
+        }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp.Tests/Components/ReferenceData/CategoriesTableTestFixture.cs:414]: Add new line at the end

File: COMETwebapp.Tests/Components/ReferenceData/CategoriesTableTestFixture.cs [hunk 2]
```diff
@@ -38,13 +38,15 @@ namespace COMETwebapp.Tests.Components.ReferenceData
...
+    using CDP4Dal.Events;
     using CDP4Dal.Permission;
 
     using COMETwebapp.Components.ReferenceData;
+    using COMETwebapp.Components.UserManagement;
     using COMETwebapp.Services.SessionManagement;
     using COMETwebapp.Tests.Helpers;
     using COMETwebapp.ViewModels.Components.ReferenceData;
-
+    using COMETwebapp.Wrappers;
```
Comment (jaimeatrhea) [inline COMETwebapp.Tests/Components/ReferenceData/CategoriesTableTestFixture.cs:49]: Separate usings

File: COMETwebapp.Tests/Components/ReferenceData/ParameterTypeTableTestFixture.cs [hunk 3]
```diff
@@ -309,4 +308,4 @@ public async Task VerifyOnInitialized()
             });
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp.Tests/Components/ReferenceData/ParameterTypeTableTestFixture.cs:311]: Add new line at the end

File: COMETwebapp.Tests/Components/SystemRepresentation/ElementDefinitionDetailsTestFixture.cs [hunk 4]
```diff
@@ -100,4 +100,4 @@ public void VerifyComponent()
             });
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp.Tests/Components/SystemRepresentation/ElementDefinitionDetailsTestFixture.cs:103]: Add new line at the end

File: COMETwebapp.Tests/Components/SystemRepresentation/SystemTreeTestFixture.cs [hunk 5]
```diff
@@ -78,4 +78,4 @@ public void VerifyComponent()
             Assert.That(renderer.Markup, Does.Contain("node1"));
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp.Tests/Components/SystemRepresentation/SystemTreeTestFixture.cs:81]: Add new line at the end

File: COMETwebapp.Tests/Components/UserManagement/UserManagementTableTestFixture.cs [hunk 6]
```diff
@@ -391,4 +391,4 @@ public async Task VerifyUnDeprecatingPerson()
             Assert.That(this.viewModel.PopupVisible, Is.False);
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp.Tests/Components/UserManagement/UserManagementTableTestFixture.cs:394]: Add new line at the end

File: COMETwebapp.Tests/Helpers/DevExpressBlazorTestHelper.cs [hunk 7]
```diff
@@ -165,4 +165,4 @@ public DateTime GetDateTimeNow()
         /// </summary>
         Task<DeviceInfo> IEnvironmentInfo.DeviceInfo => Task.FromResult(new DeviceInfo(false));
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp.Tests/Helpers/DevExpressBlazorTestHelper.cs:168]: Add new line at the end

File: COMETwebapp.Tests/Services/SessionManagement/SessionServiceTest.cs [hunk 8]
```diff
@@ -323,4 +323,4 @@ public void VerifyGetAvailableDomains()
             Assert.That(this.sessionService.Session.RetrieveSiteDirectory().Domain.ToList(), Has.Count.EqualTo(1));
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp.Tests/Services/SessionManagement/SessionServiceTest.cs:326]: Add new line at the end

File: COMETwebapp/COMETwebapp.csproj [hunk 9]
```diff
@@ -37,4 +37,4 @@
     <ItemGroup>
       <_ContentIncludedByDefault Remove="Components\SystemRepresentation\SystemRepresentationBody.razor" />
     </ItemGroup>
-</Project>
+</Project>
```
Comment (jaimeatrhea) [inline COMETwebapp/COMETwebapp.csproj:40]: Add new line at the end

File: COMETwebapp/Components/ReferenceData/CategoriesTable.razor.cs [hunk 10]
```diff
@@ -26,16 +26,16 @@ namespace COMETwebapp.Components.ReferenceData
 {
     using System.Reactive.Linq;
     using System.Threading.Tasks;
-
+    using CDP4Common.SiteDirectoryData;
+    
     using COMETwebapp.ViewModels.Components.ReferenceData;
-
+    using COMETwebapp.Wrappers;
```
Comment (jaimeatrhea) [inline COMETwebapp/Components/ReferenceData/CategoriesTable.razor.cs:32]: Separate usings

File: COMETwebapp/Components/ReferenceData/CategoriesTable.razor.cs [hunk 11]
```diff
@@ -91,12 +105,10 @@ protected override Task OnInitializedAsync()
...
-            this.Disposables.Add(this.WhenAnyValue(x => x.ViewModel.IsAllowedToWrite).Subscribe(_ => this.InvokeAsync(this.StateHasChanged)));
-
             this.Disposables.Add(this.ViewModel.Rows.CountChanged.Subscribe(_ => this.InvokeAsync(this.StateHasChanged)));
             this.Disposables.Add(this.ViewModel.Rows.Connect().AutoRefresh().Subscribe(_ => this.InvokeAsync(this.StateHasChanged)));
 
             return base.OnInitializedAsync();
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/Components/ReferenceData/CategoriesTable.razor.cs:114]: Add new line at the end

File: COMETwebapp/Components/ReferenceData/ParameterTypeTable.razor.cs [hunk 12]
```diff
@@ -92,12 +92,10 @@ protected override Task OnInitializedAsync()
...
-            this.Disposables.Add(this.WhenAnyValue(x => x.ViewModel.IsAllowedToWrite).Subscribe(_ => this.InvokeAsync(this.StateHasChanged)));
-
             this.Disposables.Add(this.ViewModel.Rows.CountChanged.Subscribe(_ => this.InvokeAsync(this.StateHasChanged)));
             this.Disposables.Add(this.ViewModel.Rows.Connect().AutoRefresh().Subscribe(_ => this.InvokeAsync(this.StateHasChanged)));
 
             return base.OnInitializedAsync();
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/Components/ReferenceData/ParameterTypeTable.razor.cs:101]: Add new line at the end

File: COMETwebapp/Components/SystemRepresentation/ElementDefinitionDetails.razor.cs [hunk 13]
```diff
@@ -55,4 +55,4 @@ protected override Task OnInitializedAsync()
             return base.OnInitializedAsync();
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/Components/SystemRepresentation/ElementDefinitionDetails.razor.cs:58]: Add new line at the end

File: COMETwebapp/Components/SystemRepresentation/SystemTree.razor.cs [hunk 14]
```diff
@@ -42,4 +42,4 @@ public partial class SystemTree
         [Parameter]
         public ISystemTreeViewModel ViewModel { get; set; }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/Components/SystemRepresentation/SystemTree.razor.cs:45]: Add new line at the end

File: COMETwebapp/Components/UserManagement/UserManagementTable.razor.cs [hunk 15]
```diff
@@ -154,4 +154,4 @@ private static void DisableDeprecatedPerson(GridCustomizeElementEventArgs e)
             }
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/Components/UserManagement/UserManagementTable.razor.cs:157]: Add new line at the end

File: COMETwebapp/Model/ApplicationCards.cs [hunk 16]
```diff
@@ -110,4 +110,4 @@ public static class ApplicationCards
             }
         };
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/Model/ApplicationCards.cs:113]: Add new line at the end

File: COMETwebapp/Model/SystemNode.cs [hunk 17]
```diff
@@ -90,4 +90,4 @@ public IReadOnlyList<SystemNode> GetChildren()
             return this.Children.AsReadOnly();
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/Model/SystemNode.cs:93]: Add new line at the end

File: COMETwebapp/Pages/SystemRepresentation/SystemRepresentation.razor [hunk 18]
```diff
@@ -1,30 +1,30 @@
...
+------------------------------------------------------------------------------->
+@using COMETwebapp.Utilities
+@inherits SingleIterationPageTemplate
+@attribute [Route(ConstantValues.SystemRepresentationPage)]
+
+<SingleIterationApplicationTemplate IterationId="@this.RequestedIteration">
+    <Body>
+        <SystemRepresentationBody />
+    </Body>
 </SingleIterationApplicationTemplate>
```
Comment (jaimeatrhea) [inline COMETwebapp/Pages/SystemRepresentation/SystemRepresentation.razor:30]: Add new line at the end

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 19]
```diff
@@ -217,5 +233,134 @@ private void UpdateRows(IEnumerable<CategoryRowViewModel> rowsToDisplay)
...
+            {
+                this.DataSource.AddRange(referenceDataLibrary.DefinedCategory);
+            }
+            this.ReferenceDataLibraries = this.sessionService.Session.RetrieveSiteDirectory().AvailableReferenceDataLibraries();
+            this.UpdateProperties(this.DataSource.Items);
+            this.RefreshAccessRight();
+        }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:366]: Add new line at the end

File: COMETwebapp/ViewModels/Components/ReferenceData/ICategoriesTableViewModel.cs [hunk 20]
```diff
@@ -59,4 +119,4 @@ public interface ICategoriesTableViewModel : IDisposable
         /// <returns>A <see cref="Task" /> representing any asynchronous operation.</returns>
         void OnInitializedAsync();
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/ICategoriesTableViewModel.cs:122]: Add new line at the end

File: COMETwebapp/ViewModels/Components/ReferenceData/IParameterTypeTableViewModel.cs [hunk 21]
```diff
@@ -58,4 +53,4 @@ public interface IParameterTypeTableViewModel: IDisposable
         /// <returns>A <see cref="Task" /> representing any asynchronous operation.</returns>
         void OnInitializedAsync();
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/IParameterTypeTableViewModel.cs:56]: Add new line at the end

File: COMETwebapp/ViewModels/Components/ReferenceData/ParameterTypeTableViewModel.cs [hunk 22]
```diff
@@ -221,4 +217,4 @@ private void UpdateRows(IEnumerable<ParameterTypeRowViewModel> rowsToDisplay)
             }
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/ParameterTypeTableViewModel.cs:220]: Add new line at the end

File: COMETwebapp/ViewModels/Components/ReferenceData/Rows/CategoryRowViewModel.cs [hunk 23]
```diff
@@ -138,4 +152,4 @@ public void UpdateProperties(CategoryRowViewModel categoryRow)
             this.IsDeprecated = categoryRow.IsDeprecated;
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/Rows/CategoryRowViewModel.cs:155]: Add new line at the end

File: COMETwebapp/ViewModels/Components/ReferenceData/Rows/ParameterTypeRowViewModel.cs [hunk 24]
```diff
@@ -169,4 +183,4 @@ public void UpdateProperties(ParameterTypeRowViewModel parameterTypeRow)
             this.IsDeprecated = parameterTypeRow.IsDeprecated;
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/Rows/ParameterTypeRowViewModel.cs:186]: Add new line at the end

File: COMETwebapp/ViewModels/Components/UserManagement/IUserManagementTableViewModel.cs [hunk 25]
```diff
@@ -149,4 +149,4 @@ public interface IUserManagementTableViewModel : IDisposable
         /// <returns>A <see cref="Task" /> representing any asynchronous operation.</returns>
         void OnInitializedAsync();
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/UserManagement/IUserManagementTableViewModel.cs:152]: Add new line at the end

File: COMETwebapp/ViewModels/Components/UserManagement/Rows/PersonRowViewModel.cs [hunk 26]
```diff
@@ -166,4 +166,4 @@ public void UpdateProperties(PersonRowViewModel person)
             this.IsDeprecated = person.IsDeprecated;
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/UserManagement/Rows/PersonRowViewModel.cs:169]: Add new line at the end

File: COMETwebapp/ViewModels/Components/UserManagement/UserManagementTableViewModel.cs [hunk 27]
```diff
@@ -363,4 +363,4 @@ private void UpdateRows(IEnumerable<PersonRowViewModel> rowsToDisplay)
             }
         }
     }
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/ViewModels/Components/UserManagement/UserManagementTableViewModel.cs:366]: Add new line at the end

File: COMETwebapp/_Imports.razor [hunk 28]
```diff
@@ -35,4 +35,4 @@
 @using COMETwebapp.Components.Viewer.PopUps
 @using COMETwebapp.Components.Viewer.PropertiesPanel
 @using BlazorStrap
-@using DevExpress.Blazor
+@using DevExpress.Blazor
```
Comment (jaimeatrhea) [inline COMETwebapp/_Imports.razor:38]: Add new line at the end

File: COMETwebapp/wwwroot/css/app.css [hunk 29]
```diff
@@ -408,4 +408,4 @@ html, body, #app {
 
 .highlighted-item {
     background-color: #DEDFE4;
-}
+}
```
Comment (jaimeatrhea) [inline COMETwebapp/wwwroot/css/app.css:411]: Add new line at the end

File: COMETwebapp/wwwroot/index.html [hunk 30]
```diff
@@ -75,4 +75,4 @@
     
 </body>
 
-</html>
+</html>
```
Comment (jaimeatrhea) [inline COMETwebapp/wwwroot/index.html:78]: Add new line at the end

**Review by jaimeatrhea — CHANGES_REQUESTED** [review]

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 31]
```diff
@@ -111,31 +114,39 @@ public CategoriesTableViewModel(ISessionService sessionService)
...
+
+        /// <summary>
+        ///  Indicates if confirmation popup is visible
+        /// </summary>
+        public bool popupVisible { get; set; } = false;
+
+        /// <summary>
+        ///     popum message dialog
+        /// </summary>
+        public string popupDialog { get; set; }
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:144]: Rename this property to something coherent with the previous comment. Also, any public property should start with an uppercase

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 32]
```diff
@@ -217,5 +233,134 @@ private void UpdateRows(IEnumerable<CategoryRowViewModel> rowsToDisplay)
...
+            {
+                await this.DeprecatingCategory();
+            }
+            popupVisible = false;
+        }
+
+        /// <summary>
+        /// Action invoked when the deprecate or undeprecate button is clicked
+        /// </summary>
+        public void OnDeprecateUnDeprecateButtonClick(GridCommandColumnCellDisplayTemplateContext context)
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:264]: Missing documentation. Also, provide a CategoryRowViewModel as parameter and not the GridCommandColumnCellDisplay. That type is DevExpress related, VM should not be dependent of UI framework

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 33]
```diff
@@ -217,5 +233,134 @@ private void UpdateRows(IEnumerable<CategoryRowViewModel> rowsToDisplay)
...
+        /// <returns>A <see cref="Task" /></returns>
+        public async Task AddingCategory()
+        {
+            var thingsToCreate = new List<Thing>();
+
+            if (this.SelectedSuperCategories.Any())
+            {
+                this.Category.SuperCategory = this.SelectedSuperCategories.ToList();
+            }
+            if (this.SelectedPermissibleClasses.Any())
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:329]: missing break line

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 34]
```diff
@@ -217,5 +233,134 @@ private void UpdateRows(IEnumerable<CategoryRowViewModel> rowsToDisplay)
...
+            {
+                this.Category.PermissibleClass = this.SelectedPermissibleClasses.Select(x => x.ClassKind).ToList();
+            }
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:332]: missing break line

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 35]
```diff
@@ -217,5 +233,134 @@ private void UpdateRows(IEnumerable<CategoryRowViewModel> rowsToDisplay)
...
+            this.Category.Container = this.SelectedReferenceDataLibrary;
+            thingsToCreate.Add(this.Category);
+            var clonedRDL = this.SelectedReferenceDataLibrary.Clone(false);
+            clonedRDL.DefinedCategory.Add(this.Category);
+            try
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:337]: missing break line

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 36]
```diff
@@ -217,5 +233,134 @@ private void UpdateRows(IEnumerable<CategoryRowViewModel> rowsToDisplay)
...
+        ///     Override this method if you will perform an asynchronous operation and
+        ///     want the component to refresh when that operation is completed.
+        /// </summary>
+        /// <returns>A <see cref="Task" /> representing any asynchronous operation.</returns>
+        public void OnInitializedAsync()
+        {
+            foreach (var referenceDataLibrary in this.sessionService.Session.RetrieveSiteDirectory().AvailableReferenceDataLibraries())
+            {
+                this.DataSource.AddRange(referenceDataLibrary.DefinedCategory);
+            }
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:360]: missing break line

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 37]
```diff
@@ -217,5 +233,134 @@ private void UpdateRows(IEnumerable<CategoryRowViewModel> rowsToDisplay)
...
+        }
+
+        /// <summary>
+        ///     Method invoked when the component is ready to start, having received its
+        ///     initial parameters from its parent in the render tree.
+        ///     Override this method if you will perform an asynchronous operation and
+        ///     want the component to refresh when that operation is completed.
+        /// </summary>
+        /// <returns>A <see cref="Task" /> representing any asynchronous operation.</returns>
+        public void OnInitializedAsync()
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:355]: this method is not Async. Remove the Async on the method name and remove the return on documentation

File: COMETwebapp/ViewModels/Components/ReferenceData/ICategoriesTableViewModel.cs [hunk 38]
```diff
@@ -24,9 +24,13 @@
 
 namespace COMETwebapp.ViewModels.Components.ReferenceData
 {
+    using CDP4Common.CommonData;
     using CDP4Common.SiteDirectoryData;
 
     using COMETwebapp.ViewModels.Components.ReferenceData.Rows;
+    using COMETwebapp.Wrappers;
+
+    using DevExpress.Blazor;
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/ICategoriesTableViewModel.cs:33]: VM should not depend on UI Framework

**Review by antoineatrhea — CHANGES_REQUESTED** [review]
Small changes needed

File: COMETwebapp/Extensions/EnumeratorExtension.cs [hunk 39]
```diff
@@ -0,0 +1,64 @@
...
+        private static readonly Dictionary<Enum, string> displayName = new();
+
+        /// <summary>
+        ///     Retrieve the <see cref="DisplayAttribute" /> of an <see cref="Enum" /> value
+        /// </summary>
+        /// <param name="value">The <see cref="Enum" /> value</param>
+        /// <returns>The <see cref="DisplayAttribute" /> value if found</returns>
+        public static string GetEnumDisplayName(this Enum value)
+        {
+            if (displayName.ContainsKey(value))
```
[the remark this answers is not in the document]
Comment (antoineatrhea) [reply COMETwebapp/Extensions/EnumeratorExtension.cs]: Saw that it has been fixed on new commit, discard this comment

File: COMETwebapp/Components/ReferenceData/CategoriesTable.razor.cs [hunk 40]
```diff
@@ -80,6 +83,20 @@ private static void DisableDeprecatedCategories(GridCustomizeElementEventArgs e)
...
 
+        /// <summary>
+        ///     Method invoked when creating a new category
+        /// </summary>
+        /// <param name="e">A <see cref="GridCustomizeEditModelEventArgs"/>
+        private void CustomizeEditCategory(GridCustomizeEditModelEventArgs e)
+        {
+            var dataItem = (Category)e.DataItem;
+            if (dataItem == null)
+                e.EditModel = new Category { };
```
Comment (antoineatrhea) [inline COMETwebapp/Components/ReferenceData/CategoriesTable.razor.cs:94]: please encapsulate the line inside {}

**Review by antoineatrhea — APPROVED** [review]
One small thing spotted but otherwise, ok for me

**Review by jaimeatrhea — APPROVED** [review]

**Pull request merged**